data_model/src/events/data/events.rs

```diff
@@ -0,0 +1,301 @@
+
+    use super::*;
+
+    /// Domain Event
+    #[derive(Clone, PartialEq, Eq, Debug, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    #[non_exhaustive]
+    #[allow(missing_docs)]
+    pub enum DomainEvent {
+        Account(AccountEvent),
+        AssetDefinition(AssetDefinitionEvent),
+        Created(DomainId),
+        Deleted(DomainId),
+        MetadataInserted(DomainId),
+        MetadataRemoved(DomainId),
+    }
+
+    impl Identifiable for DomainEvent {
+        type Id = DomainId;
+    }
+
+    impl IdTrait for DomainEvent {
+        fn id(&self) -> &DomainId {
+            match self {
+                Self::Account(account) => &account.id().domain_id,
+                Self::AssetDefinition(asset_definition) => &asset_definition.id().domain_id,
+                Self::Created(id)
+                | Self::Deleted(id)
+                | Self::MetadataInserted(id)
+                | Self::MetadataRemoved(id) => id,
+            }
+        }
+    }
+}
+
+mod trigger {
+    //! This module contains `TriggerEvent` and its impls
+
+    use super::*;
+
+    /// Trigger Event
+    #[derive(Clone, PartialEq, Eq, Debug, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    #[non_exhaustive]
+    #[allow(missing_docs)]
+    pub enum TriggerEvent {
+        Created(TriggerId),
+        Deleted(TriggerId),
+        Extended(TriggerId),
+        Shortened(TriggerId),
+    }
+
+    impl Identifiable for TriggerEvent {
+        type Id = TriggerId;
+    }
+
+    impl IdTrait for TriggerEvent {
+        fn id(&self) -> &TriggerId {
+            match self {
+                Self::Created(id)
+                | Self::Deleted(id)
+                | Self::Extended(id)
+                | Self::Shortened(id) => id,
+            }
+        }
+    }
+
+    impl From<Register<Trigger>> for DataEvent {
+        fn from(src: Register<Trigger>) -> Self {
+            Self::Trigger(TriggerEvent::Created(src.object.id))
+        }
+    }
+
+    impl From<Unregister<Trigger>> for DataEvent {
+        fn from(src: Unregister<Trigger>) -> Self {
+            Self::Trigger(TriggerEvent::Deleted(src.object_id))
+        }
+    }
+
+    impl From<Mint<Trigger, u32>> for DataEvent {
+        fn from(src: Mint<Trigger, u32>) -> Self {
+            Self::Trigger(TriggerEvent::Extended(src.destination_id))
+        }
+    }
+
+    impl From<Burn<Trigger, u32>> for DataEvent {
+        fn from(src: Burn<Trigger, u32>) -> Self {
+            Self::Trigger(TriggerEvent::Shortened(src.destination_id))
+        }
+    }
+}
+
+/// World event
+///
+/// Does not participate in `Event`, but useful for events warranties when modifying `wsv`
+#[derive(
+    Clone, PartialEq, Eq, Debug, Decode, Encode, Deserialize, Serialize, FromVariant, IntoSchema,
+)]
+#[allow(missing_docs)]
+pub enum WorldEvent {
+    Domain(domain::DomainEvent),
+    Peer(peer::PeerEvent),
+
+    #[cfg(feature = "roles")]
+    Role(role::RoleEvent),
+}
```

Comment:
I dislike repacking events into `WorldEvent` because concept of the world is internal to the blockchain, it has no meaning for the user inspecting events.

I dislike repacking events in other places as well(like `DomainEvent::Account` and `Account::Asset`, but I'm still not able to make a strong argument against it. I think these events should be able to stand on their own

Reply:
`WorldEvent` does not participate in main `Event` structure. It is used for `modify_` methods API. User should look only for `Event` structure and its internal fields.

Packing events like `DomainEvent::Account` allows us not only handy `modify_` methods API, but also possibility for user to listen for any  `AccountEvent` happened with Domain with particular id

Reply:
> Packing events like DomainEvent::Account allows us not only handy modify_ methods API, but also possibility for user to listen for any AccountEvent happened with Domain with particular id

I think this is orthogonal to the layout of events. Listening to events can be(and is intended to be) controlled with filtering with the same effect.

I can see that you are pushing events up to the `modify_world` where you finally publish them. This is why you need `WorldEvent`/`Domain::Account`/`Account::Asset`. Alternative would be to publish event directly in `modify_domain`/`modify_account`/`modify_asset`.

Reply:
I think this architecture needs iterative improvement. Strong typing can prevent a class of bugs, and it's good-enough for now. Since I want to isolate the World from the rest of core, it might not be such a bad idea.